### PR TITLE
Refactor simple_embedding_run with separate create_sample_device()

### PR DESCRIPTION
**Pull request by hcindyl**
simple_embedding_run of different backend targets shares common
code struction except how the device is created. Refactor the code with
simple_embedding_backend = {simple_embedding, device_backend} so the
common interface/code can be shared.

### CODE REVIEW

File: iree/samples/simple_embedding/device_local_sync.c [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+// A example of setting up the dylib-sync driver.
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/hal/local/executable_loader.h"
+#include "iree/hal/local/loaders/legacy_library_loader.h"
+#include "iree/hal/local/sync_driver.h"
+
+iree_status_t create_sample_device(iree_hal_device_t** device) {
+  // Set paramters for the device created in the next ste.
```
Comment (marbre) [inline iree/samples/simple_embedding/device_local_sync.c:24]:
```suggestion
  // Set parameters for the device created in the next step.
```
NIT: Those typos were unfortunately introduced by myself.

File: iree/samples/simple_embedding/device_vulkan.c [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+// limitations under the License.
+
+// A example of setting up the the vulkan driver.
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/hal/vulkan/registration/driver_module.h"
+
+iree_status_t create_sample_device(iree_hal_device_t** device) {
+  // Only register vulkan HAL driver
```
Comment (marbre) [inline iree/samples/simple_embedding/device_vulkan.c:22]:
```suggestion
  // Only register vulkan HAL driver.
```

File: iree/samples/simple_embedding/device_dylib.c [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+// limitations under the License.
+
+// A example of setting up the the dylib driver.
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/hal/dylib/registration/driver_module.h"
+
+iree_status_t create_sample_device(iree_hal_device_t** device) {
+  // Only register dylib HAL driver
```
Comment (marbre) [inline iree/samples/simple_embedding/device_dylib.c:22]:
```suggestion
  // Only register dylib HAL driver.
```

**Review by marbre — CHANGES_REQUESTED** [review]
Only some small nits. However, I don't know if we really want to drop demonstrate how to use `iree_hal_register_all_available_drivers()`, which was used in `simple_embedding_run.c`.

File: iree/samples/simple_embedding/device_dylib.c (hunk 3 above)
Comment (hcindyl) [reply iree/samples/simple_embedding/device_dylib.c]: Done.

File: iree/samples/simple_embedding/device_local_sync.c (hunk 1 above)
Comment (hcindyl) [reply iree/samples/simple_embedding/device_local_sync.c]: Done.

**Review by marbre — APPROVED** [review]
LGTM!

File: iree/samples/simple_embedding/BUILD [hunk 4]
```diff
@@ -115,17 +139,15 @@ endif()
 
 run_binary_test(
     name = "simple_embedding_dylib_test",
-    args = ["dylib"],
-    test_binary = ":simple_embedding_run",
+    test_binary = ":simple_embedding_dylib",
```
Comment (GMNGeoffrey) [inline iree/samples/simple_embedding/BUILD:142]: Nit: while you're here, could you add the appropriate driver tags to each of these tests?

"driver=vulkan", "driver=dylib", etc.

File: iree/samples/simple_embedding/simple_embedding.c [hunk 5]
```diff
@@ -12,15 +12,12 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// A example of setting up the HAL mddule to run simple pointwise array
-// multiplication with the dylib-sync driver.
+// A example of setting up the HAL module to run simple pointwise array
+// multiplication with the dylib driver.
```
Comment (ScottTodd) [inline iree/samples/simple_embedding/simple_embedding.c:16]: reword, this isn't just for the dylib driver now

File: iree/samples/simple_embedding/simple_embedding.c [hunk 6]
```diff
@@ -32,6 +29,10 @@
 #include "iree/samples/simple_embedding/simple_embedding_test_bytecode_module_c.h"
 #endif
 
+// A function to create the HAL device from the different backend targets.
+// The HAL device is returned based on the implementation.
+extern iree_status_t create_sample_device(iree_hal_device_t** device);
```
Comment (ScottTodd) [inline iree/samples/simple_embedding/simple_embedding.c:34]: could document ownership here (the device returned must be released by the caller)

File: iree/samples/simple_embedding/BUILD (hunk 4 above)
Comment (GMNGeoffrey) [reply iree/samples/simple_embedding/BUILD:142]:
```suggestion
    test_binary = ":simple_embedding_dylib",
    tags = ["driver=dylib"],
```

**Review by ScottTodd — APPROVED** [review]

File: iree/samples/simple_embedding/BUILD (hunk 4 above)
Comment (hcindyl) [reply iree/samples/simple_embedding/BUILD:142]: run_binary_test doesn't support the tags field yet. :( https://github.com/google/iree/blob/7c2d310/build_tools/bazel_to_cmake/bazel_to_cmake_converter.py#L703

Can add this in a following PR.

File: iree/samples/simple_embedding/simple_embedding.c (hunk 6 above)
Comment (hcindyl) [reply iree/samples/simple_embedding/simple_embedding.c]: Done.

File: iree/samples/simple_embedding/simple_embedding.c (hunk 5 above)
Comment (hcindyl) [reply iree/samples/simple_embedding/simple_embedding.c]: Done.

**Pull request merged**